File: src/the_simplest_demo.cpp
/*
@description:
This node publishes velocities (in open-loop) to teleoperate a the simulated drone, and it subscribes to the drone 
odometry in order to obtain its position and its orientation.
Both the translational velocities (Vx, Vy, Vz) and the rotational velocity (Wz) are published on the "/r1/cmd_vel" topic.
Both the position (x,y,z) and the orientation (quaternion) are received from the "/r1/odom" topic, and they are displayed.

-- Use ctrl - C to finish this node. Important: drone will keep moving

DIRECTIONS:
In different terminals:
$ ros2 launch my_uavs ign_world_launch.py
	# Launchs one simulated quadrotor in Ignition and runs ros_ign_bridge

$ ros2 run my_uavs the_simplest_demo
	# Runs this node
*/

#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>

using namespace std::chrono_literals;
using std::placeholders::_1;

#define RAD2DEG 57.295779513


class SimpleMovement : public rclcpp::Node
{
  public:
    SimpleMovement() : Node("the_simplest_demo")
    {
      publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/r1/cmd_vel", 2);

      timer_ = this->create_wall_timer(50ms, std::bind(&SimpleMovement::loop, this)); //50ms = 20 Hz
      
      subscriber_ = this->create_subscription<nav_msgs::msg::Odometry>("/r1/odom", 
            2, std::bind(&SimpleMovement::odom_callback, this, _1));

      RCLCPP_WARN(this->get_logger(), "Node has been initialized");
    }
    
    ~SimpleMovement() //Class destructor definition
    {
      RCLCPP_INFO(this->get_logger(), "Node has been terminated"); 
    }
    
    void odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
    {
      x_ = msg->pose.pose.position.x;
      y_ = msg->pose.pose.position.y;
      z_ = msg->pose.pose.position.z;
      
      //GET THE EULER ANGLES FROM QUATERNION
      tf2::Quaternion q( msg->pose.pose.orientation.x, msg->pose.pose.orientation.y,
                    msg->pose.pose.orientation.z, msg->pose.pose.orientation.w);

      // 3x3 Rotation matrix from quaternion
      tf2::Matrix3x3 m(q);

      // Roll, Pitch and Yaw from rotation matrix
      double roll, pitch, yaw;

      m.getRPY(roll, pitch, yaw); //Euler angles in rad
      roll_deg_ = roll * RAD2DEG;
      pitch_deg_ = pitch * RAD2DEG;
      yaw_deg_ = yaw * RAD2DEG;
    }

    void loop()
    {
      vel_msg_.linear.x = 0.4;
      vel_msg_.linear.y = 0.0;
      vel_msg_.linear.z = 0.1;
      vel_msg_.angular.z = 0.6;
      
      //Display the drone pose using a frequency divisor
      if(freq_div_ == 20)
      {
        RCLCPP_INFO(this->get_logger(), "x: %.2f, y: %.2f, z: %.2f, R: %.2f, P: %.2f, Y: %.2f", 
            x_,y_,z_, roll_deg_,pitch_deg_,yaw_deg_);
        freq_div_ = 0;
      }else freq_div_++;

      publisher_->publish(vel_msg_); //Publish velocity
    }

  private:
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr subscriber_;
   
    geometry_msgs::msg::Twist vel_msg_; //Create a msg obj to publish velocity
    
    int freq_div_;
    double x_, y_, z_, roll_deg_, pitch_deg_, yaw_deg_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleMovement>();
  rclcpp::spin(node);

  rclcpp::shutdown();
  return 0;
}

File: scripts/the_simplest_demo.py
#!/usr/bin/env python3

"""
@description:
This node publishes velocities (in open-loop) to teleoperate a the simulated drone, and it subscribes to the drone 
odometry in order to obtain its position and its orientation.
Both the translational velocities (Vx, Vy, Vz) and the rotational velocity (Wz) are published on the "/r1/cmd_vel" topic.
Both the position (x,y,z) and the orientation (quaternion) are received from the "/r1/odom" topic, and they are displayed.

DIRECTIONS:
In different terminals:
$ ros2 launch my_uavs ign_world_launch.py
	# Launchs one simulated quadrotor in Ignition and runs ros_ign_bridge

$ ros2 run my_uavs the_simplest_demo.py
	# Runs this node (python script)
    # To give execution permission$ chmod +x the_simplest_demo.py

-- To finish this node, please press '[Ctrl]+C'. Important: drone will keep moving
"""

__author__ = "David Caceres"
__contact__ = "davidxcaceres1@gmail.com"

import rclpy#, math
from rclpy.node import Node
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
#To work with Euler angles and quaternions conversions
import tf_transformations

"""
It is required to install:
$ sudo apt install ros-$ROS_DISTRO-tf-transformations
$ sudo apt install python3-pip
$ sudo pip3 install transforms3d #Python package
"""

RAD2DEG = 57.295779513


class SimpleMovement(Node):

    def __init__(self):
        super().__init__('the_simplest_demo') # class’s constructor

        self.publisher = self.create_publisher(Twist, '/r1/cmd_vel', 2)

        self.timer = self.create_timer(0.05, self.loop) # 50ms = 20 Hz

        self.subscription = self.create_subscription(Odometry,'/r1/odom',self.odom_callback, 2)

        self.get_logger().warn('Node has been initialized')

        self.vel_msg = Twist()
        self.freq_div = 0
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0
        self.roll_deg = 0.0
        self.pitch_deg = 0.0
        self.yaw_deg = 0.0



    def __del__(self):
        self.get_logger().info('Node has been terminated')
        

    def odom_callback(self, msg):
        self.x = msg.pose.pose.position.x
        self.y = msg.pose.pose.position.y
        self.z = msg.pose.pose.position.z

        #GET THE EULER ANGLES FROM QUATERNION
        q = [0.0,0.0,0.0,1.0] #Define a unitary quaternion

        q[0] = msg.pose.pose.orientation.x
        q[1] = msg.pose.pose.orientation.y
        q[2] = msg.pose.pose.orientation.z
        q[3] = msg.pose.pose.orientation.w

        r, p, y = tf_transformations.euler_from_quaternion(q)
        self.roll_deg = r * RAD2DEG
        self.pitch_deg = p * RAD2DEG
        self.yaw_deg = y * RAD2DEG


    def loop(self):
        self.vel_msg.linear.x = 0.4
        self.vel_msg.linear.y = 0.0
        self.vel_msg.linear.z = 0.1
        self.vel_msg.angular.z = 0.6

        # Display the drone pose using a frequency divisor
        if self.freq_div == 20:
            self.get_logger().info('x: %.2f, y: %.2f, z: %.2f, R: %.2f, P: %.2f, Y: %.2f'
                    % (self.x,self.y,self.z, self.roll_deg, self.pitch_deg, self.yaw_deg) )
            self.freq_div = 0
        else:
            self.freq_div = self.freq_div +1

        self.publisher.publish(self.vel_msg) # Publish velocity


def main(args=None):
    rclpy.init(args=args)
    node = SimpleMovement()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        print('Ctrl-C detected, signal_handler(signum=2)')
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

